packages/core/src/core/diff.ts

```diff
@@ -13,137 +13,139 @@
     selectors,
     source,
     referenceSourcesIndex,
-  })).toArray()
+  })).toArray(), 'diff.getFilteredIds')
 )
 
+const createMatchers = async (
+  beforeElementsSrc: elementSource.ElementsSource,
+  afterElementsSrc: elementSource.ElementsSource,
+  referenceSourcesIndex: remoteMap.ReadOnlyRemoteMap<ElemID[]>,
+  elementSelectors: ElementSelector[]
+): Promise<{
+  isChangeMatchSelectors: (change: DetailedChange) => boolean
+  isTopLevelElementMatchSelectors: (elemID: ElemID) => boolean
+}> => {
+  const beforeMatchingElemIDs = await getFilteredIds(
+    elementSelectors,
+    beforeElementsSrc,
+    referenceSourcesIndex,
+  )
+  const afterMatchingElemIDs = await getFilteredIds(
+    elementSelectors,
+    afterElementsSrc,
+    referenceSourcesIndex,
+  )
+
+  const allMatchingTopLevelElemIDsSet = new Set<string>(
+    beforeMatchingElemIDs.concat(afterMatchingElemIDs)
+      .map(id => id.createTopLevelParentID().parent.getFullName())
+  )
+  const isTopLevelElementMatchSelectors = (elemID: ElemID): boolean =>
+    allMatchingTopLevelElemIDsSet.has(elemID.getFullName())
+
+  // skipping change matching filtering when all selectors are top level
+  if (beforeMatchingElemIDs.every(id => id.isTopLevel())
+    && afterMatchingElemIDs.every(id => id.isTopLevel())) {
+    log.debug('all selectors are top level. skipping change matching filtering')
+    return {
+      isTopLevelElementMatchSelectors,
+      isChangeMatchSelectors: () => true,
+    }
+  }
+
+  // this set will be used to look for a change's id or its parents
+  const beforeMatchingElemIDsSet = new Set(beforeMatchingElemIDs
+    .map(elemId => elemId.getFullName()))
+  const afterMatchingElemIDsSet = new Set(afterMatchingElemIDs
+    .map(elemId => elemId.getFullName()))
+
+  // this set will be used to check if a change's id is a parent of a selector-matched elemID
+  const beforeAllParentsMatchingElemIDsSet = new Set(beforeMatchingElemIDs
+    .flatMap(createAllElemIdParents)
+    .map(elemId => elemId.getFullName()))
+  const afterAllParentsMatchingElemIDsSet = new Set(afterMatchingElemIDs
+    .flatMap(createAllElemIdParents)
+    .map(elemId => elemId.getFullName()))
+
+  const isChangeIdEqualOrChildOfMatchingElemId = (change: DetailedChange): boolean => {
+    const matchingElemIDsSet = isRemovalChange(change)
+      ? beforeMatchingElemIDsSet
+      : afterMatchingElemIDsSet
+    return createAllElemIdParents(change.id)
+      .some(elemId => matchingElemIDsSet.has(elemId.getFullName()))
+  }
+
+  const isChangeIdParentOfMatchingElemId = (change: DetailedChange): boolean => {
+    const allParentsMatchingElemIDsSet = isRemovalChange(change)
+      ? beforeAllParentsMatchingElemIDsSet
+      : afterAllParentsMatchingElemIDsSet
+    return allParentsMatchingElemIDsSet.has(change.id.getFullName())
+  }
+
+  const isChangeMatchSelectors = (change: DetailedChange): boolean =>
+    isChangeIdEqualOrChildOfMatchingElemId(change) || isChangeIdParentOfMatchingElemId(change)
```

Comment:
performance nit:
I think `isChangeIdParentOfMatchingElemId` would be faster than `isChangeIdEqualOrChildOfMatchingElemId` (because the former doesn't have to create all the parent IDs of the change), so it might be better to check it first
```suggestion
    isChangeIdParentOfMatchingElemId(change) || isChangeIdEqualOrChildOfMatchingElemId(change)
```